### ISSUE TITLE
Predictive mixing is broken since #226

### ISSUE THREAD

**Issue by annethoul** [MESAHub/mesa#286]
After Bill's commit 226 (April 2021) that changed 101 files, Predictive mixing does not work anymore. Whether it is used or not leads to the same wrong evolution on the CHeB when using the Ledoux criterion.

**Comment by rjfarmer** [comment]
Commit  c6592ca

Is this a: it no longer does anything whether on/off? Or it does the wrong thing when on?

**Comment by evbauer** [comment]
Anne's plot from slack seems to show that it just isn't doing anything after #226.
![image](https://user-images.githubusercontent.com/18405113/122991667-4ce97480-d373-11eb-8308-a87d585e29ae.png)

**Comment by adamjermyn** [comment]
Is it supposed to look like the red/yellow curves? Is 226 just showing
nothing at all? (a bit hard to tell on the plot)

-Adam

**Comment by evbauer** [comment]
226 is the black curve showing a roughly constant Mcore around 0.10. It's supposed to grow like red or yellow.

**Comment by adamjermyn** [comment]
If I want to look into this is `hb_2M` the case to look at?

Related 1: Is convective premixing still working, or is this purely an issue with predictive mixing?
Related 2: If the former is working, do we want to keep predictive mixing around at all? (My understanding was that convective premixing is a strict upgrade relative to predictive mixing...)

**Comment by fxt44** [comment]
> keep predictive mixing around at all?

yes! it is currently the only formalism that prevents (random) breathing pulses. this is mission critical if one attempts to say something precise about, say, the mass fraction profiles of white dwarfs.

**Comment by adamjermyn** [comment]
Ok, so we do genuinely need to fix this one. Just making sure :-)

**Comment by annethoul** [comment]
Yes, this needs to be fixed! I will look at convective premixing to see if it is also broken or not and will let you know. But it is not a strict upgrade of predictive mixing, as Frank says.

**Comment by wmwolf** [comment]
My understanding is that predictive mixing is also best for simulations that take short timesteps. Convective premixing does unphysical things at the peak of nova bursts (and likely other similar scenarios) since it advances the convective boundary and mixes far faster than is physically possible.

**Comment by annethoul** [comment]
Agree with Bill. Convective premixing assumes instantaneous mixing. Predictive mixing does not.

**Comment by adamjermyn** [comment]
I can reproduce this on current head of `main`. Looking into it.

**Comment by adamjermyn** [comment]
It's definitely on. I turned on debugging output and it certainly prints things:
```
 Predictive mixing: i, k_a, k_b, q_a, q_b, superad_min=           1        2512        2476   5.0477468908206591E-002   5.0556394414585926E-002  0.62840621392012275     
 add_predictive_mixing; model, n_conv_bdy=         140           3
 Predictive mixing at convective boundary: i, j=           1           1
   s%predictive_zone_type=any
   s%predictive_zone_loc=core
   s%predictive_bdy_loc=any
 Predictive mixing: i, j, q_top, q_bot:           1           1   5.0477468908206757E-002   0.0000000000000000     
 Exiting predictive search due to abundance reversal
 Predictive mixing: i, k_a, k_b, q_a, q_b, superad_min=           1        2530        2494   5.0477468908206757E-002   5.0556394414586092E-002  0.62840621392336260     
 add_predictive_mixing; model, n_conv_bdy=         141           3
 Predictive mixing at convective boundary: i, j=           1           1
   s%predictive_zone_type=any
   s%predictive_zone_loc=core
   s%predictive_bdy_loc=any
```

Does anyone know what the debug output means? :P

**Comment by evbauer** [comment]
I think I've found the issue. Explanation incoming...

**Comment by evbauer** [comment]
Predictive mixing needs to have `gradL_composition_term` set to 0 in its calls to MLT. The merge in #226 seems to have inadvertently undone that, and so predictive mix doesn't work. Should be an easy fix. I'll commit soon.

**Comment by adamjermyn** [comment]
Why does it need that?

**Comment by evbauer** [comment]
I think uniform composition is a fundamental assumption of the predictive mixing iterations. See section 2.1 in MESA IV.

**Comment by adamjermyn** [comment]
Oh I see! The assumption is that the composition must be uniform *in the
convection zone*. And that means that calls to MLT have to assume this when
predicting the mixing (because otherwise they'll say 'wait this is stable,
vc=0'. Right?

-Adam

**Comment by evbauer** [comment]
Right. I'm reading the old version of the code now to make sure I can get things restored properly.

**Comment by adamjermyn** [comment]
I love your commit (`hb_2m should know that it's failing`).

**Comment by evbauer** [comment]
Hopefully it will be shamed into compliance. (And thanks to Josiah for doing all the hard work setting up that infrastructure so that hopefully I just needed to put reasonable numbers in the inlist.)

**Comment by evbauer** [comment]
Fixed in 226cc95

For completeness, I should say that there are two MLT calls in `predictive_mix.f90`, but the one in the `restore_cell_loop` actually is meant to allow `gradL_composition_term` to be set. So that one was functioning properly, and only one of the MLT calls needed to be changed.
